### PR TITLE
Add CentOS on Travis CI.

### PR DESCRIPTION
**Pull request by junaruga**
Related: https://github.com/brianmario/mysql2/issues/988 .

### CODE REVIEW

File: spec/mysql2/client_spec.rb [hunk 1]
```diff
@@ -135,15 +136,22 @@ def connect(*args)
 
     # You may need to adjust the lines below to match your SSL certificate paths
     ssl_client = nil
+    option_overrides = {
```
Comment (junaruga) [inline spec/mysql2/client_spec.rb]: This modification is to pass the test on the CentOS environment that does not have `/etc/mysql/*.pem` files.

File: spec/mysql2/result_spec.rb [hunk 2]
```diff
@@ -164,7 +164,7 @@
       expect do
         res.each_with_index do |_, i|
           # Exhaust the first result packet then trigger a timeout
-          sleep 2 if i > 0 && i % 1000 == 0
+          sleep 4 if i > 0 && i % 1000 == 0
```
Comment (junaruga) [inline spec/mysql2/result_spec.rb]: This modification is to pass the test on CentOS.

File: spec/mysql2/client_spec.rb [hunk 3]
```diff
@@ -1,5 +1,6 @@
 require 'spec_helper'
 
+# rubocop:disable Metrics/BlockLength
```
Comment (junaruga) [inline spec/mysql2/client_spec.rb]: By below modification, the length of the block is more than the limitation.
As I was not sure how to fix, I added this ignored line.

File: .travis.yml [hunk 4]
```diff
@@ -73,6 +74,12 @@ matrix:
       addons:
         hosts:
           - mysql2gem.example.com
+    - rvm: 2.4
```
Comment (junaruga) [inline .travis.yml]: This part is a little bit tricky. Actually the tested Ruby on CentOS is not Ruby 2.4 but default installed Ruby on CentOS. But I was not sure how to update `rvm: foo` to show `rvm` is not used directly on Travis.

File: .travis_centos.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+set -eux
+
+# Install dependency packages and start mysqld service.
+sh .travis_setup_centos.sh
+
+# To avoid install error.
+sed -i '/eventmachine/d' Gemfile
```
Comment (simi) [inline .travis_centos.sh]: Why is this needed? Can't we fix the install error?

Comment (junaruga) [reply .travis_centos.sh]: Thanks for the comment.
It is to avoid install error by `bundle install` on CentOS.
But let me check again to fix it.

Comment (junaruga) [reply .travis_centos.sh]: The reason of the error is just because `g++` command was not installed on the environment.

Comment (junaruga) [reply .travis_centos.sh]: I fixed it and rebased.

File: .travis.yml [hunk 6]
```diff
@@ -73,6 +74,12 @@ matrix:
...
+      # Mark
```
Comment (sodabrew) [inline .travis.yml]: Who is Mark?

File: .travis_Dockerfile_centos [hunk 7]
```diff
@@ -0,0 +1,6 @@
+FROM centos
```
Comment (sodabrew) [inline .travis_Dockerfile_centos]: Suggest to use tagged version with `centos:7`.

File: .travis_centos.sh [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+set -eux
+
+# Install dependency packages and start mysqld service.
+sh .travis_setup_centos.sh
```
Comment (sodabrew) [inline .travis_centos.sh]: I think this should go into the Dockerfile, so that you can benefit from layer caching of these dependencies.

File: spec/mysql2/client_spec.rb (hunk 3 above)
Comment (sodabrew) [reply spec/mysql2/client_spec.rb]: See file .rubocop_todo.yml.

But you may be right that the better approach is not to globally increase the limit, but to ignore it for this file and have a more sensible limit apply everywhere else.

**Comment by junaruga** [comment]
@sodabrew thank you for the review! I am sorry I did not noticed until now. I will check those later.

File: .travis.yml (hunk 6 above)
Comment (junaruga) [reply .travis.yml]: "Mark" is not people's name in this case. I want to describe `env: DOCKER=centos` is just used to identify this test case from Travis test page. But I am going to remove the line.

File: spec/mysql2/client_spec.rb (hunk 3 above)
Comment (junaruga) [reply spec/mysql2/client_spec.rb]: Okay, I will change current limit 850 to 860 to prevent below situation.

```
spec/mysql2/client_spec.rb:3:1: C: Metrics/BlockLength: Block has too many lines. [851/850]
```

File: .travis_Dockerfile_centos (hunk 7 above)
Comment (junaruga) [reply .travis_Dockerfile_centos]: All right. I will change it to `centos:7`.

File: .travis_centos.sh (hunk 8 above)
Comment (junaruga) [reply .travis_centos.sh]: All right. Let me move the logic to the Dockefile.

Comment (junaruga) [reply .travis_centos.sh]: There is a problem to move the `.travis_setup_centos.sh` to `Dockerfile` RUN (that is executed on `docker build`'s timing).
We can not start mysqld daemon service in the process of `docker  build`. (Can not access the service).
So, I moved only the part to install dependency to the Dockerfile RUN part.
I add the modification as another commit to see the change easily for now.
How?
Thank you for the checking.

**Comment by junaruga** [comment]
@sodabrew I modified it!

**Comment by sodabrew** [comment]
Looks good to me! Are you ready for merge, or still hacking on this PR?

**Comment by junaruga** [comment]
@sodabrew I am ready for merge! I did squash the 2 commits to 1 commit now.

**Comment by junaruga** [comment]
Thank you! :smile:
